src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -273,27 +273,39 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             if ('' !== $host || ('' !== $scheme && 'http' !== $scheme && 'https' !== $scheme)) {
                 $port = '';
                 if ('http' === $scheme && 80 !== $this->context->getHttpPort()) {
-                    $port = ':'.$this->context->getHttpPort();
+                    $port = ':' . $this->context->getHttpPort();
                 } elseif ('https' === $scheme && 443 !== $this->context->getHttpsPort()) {
-                    $port = ':'.$this->context->getHttpsPort();
+                    $port = ':' . $this->context->getHttpsPort();
                 }
 
                 $schemeAuthority = self::NETWORK_PATH === $referenceType || '' === $scheme ? '//' : "$scheme://";
-                $schemeAuthority .= $host.$port;
+                $schemeAuthority .= $host . $port;
             }
         }
 
         if (self::RELATIVE_PATH === $referenceType) {
             $url = self::getRelativePath($this->context->getPathInfo(), $url);
         } else {
-            $url = $schemeAuthority.$this->context->getBaseUrl().$url;
+            $url = $schemeAuthority . $this->context->getBaseUrl() . $url;
         }
 
         // add a query string if needed
         $extra = array_udiff_assoc(array_diff_key($parameters, $variables), $defaults, function ($a, $b) {
             return $a == $b ? 0 : 1;
         });
 
+        // force string for an object. See https://bugs.php.net/bug.php?id=66966
+        foreach ($extra as $paramName => $param) {
+            if (!\is_object($param)) {
+                continue;
+            }
+            if (!\is_callable([$param, '__toString'])) {
+                $message = 'Cannot convert parameter "{parameter}" to string.';
+                throw new InvalidParameterException(strtr($message, '{parameter}', $paramName));
```

Comment:
Could we have the message saying something like: 

`$x = get_debug_type($param)`
"Error when generating an URL for "$name": Cannot convert parameter "$paramName" to string, "$x"  given. 

Or is this overkill?

(obviously with nice style, correct grammar and `sprintf`)